### ISSUE TITLE
Picking stops working with more than one layer

### ISSUE THREAD

**Issue by bclinkinbeard** [uber/deck.gl#3263]
I have a custom subclass of `ScatterplotLayer` that just adds my time uniforms. If I set `autoHighlight` and `pickable` to true everything works as expected and mousing over a circle highlights it. If I add a second instance, however, interactivity stops working. I can see both circles rendered on top of one another but neither highlights on mouse over. This even happens if I set one of the layers to `visible: false`, which seems especially confusing.

Any advice?

**Comment by ibgreen** [comment]
As a first sanity check, you are giving those two layer instances different `id` props right? 

The deck.gl update cycle cannot "match" multiple layers of the same type without that hint...

**Comment by bclinkinbeard** [comment]
Yep, confirmed visually and by deck debug logging it is updating attributes for two layers.

**Comment by bclinkinbeard** [comment]
Is there anything else that could be causing this? There are very few moving pieces so I'm pretty confused about where the issue could lie.

**Comment by ibgreen** [comment]
@1chandu Have we had any issues with GPU highlighting when using two layer instances of the same type? Maybe we could do a quick test to confirm?

**Comment by 1chandu** [comment]
I just verified locally with two `ScatterplotLayer`s, and don't see any issues with highlighting. Able to pick and highlight both layers's instances.

@bclinkinbeard , are you supplying your own shaders or overwriting `getShaders()` function of the `ScatterplotLayer` ?  Can you share your customization or copy paste the code here?

**Comment by bclinkinbeard** [comment]
Thanks @1chandu, I am overwriting `getShaders`, which I've included below.

```
getShaders() {
  const shaders = super.getShaders();
  shaders.fs = fs; // my custom shader
  shaders.inject = {
    'vs:#decl': `\
      attribute float instanceTimestamps;
      varying float vTimestamp;
      uniform float minTime;
      uniform float maxTime;
    `,
    'vs:#main-start': `vTimestamp = instanceTimestamps;`,
  };
  return shaders;
}
```

**Comment by 1chandu** [comment]
deck.gl` autohighliging` is GPU Accelerated and all computations happen on GPU. Fragment and Vertex shaders of all layers have code to support this.

I guess, in your custom pixel shader you don't have these instructions, copy paste line 48 to 52 from original shader `https://github.com/uber/deck.gl/blob/466c384bc40c50b3dd6749dd9f1519f72b3f7c78/modules/layers/src/scatterplot-layer/scatterplot-layer-fragment.glsl.js#L48`
Best approach here is to use `inject` to customize FS also.

**Comment by bclinkinbeard** [comment]
I already have those lines from the original shader, I only added a check of my custom uniforms. I probably could and should use inject for the FS too, but if I have those lines then it should be working, right? And it works with a single layer. Any ideas on how I could be messing up when adding multiple layers?

**Comment by 1chandu** [comment]
I am assuming:
- When you say, two layer instances, those two are your custom layer instances.
- When it is working with one layer, that is also your custom layer.

> if I have those lines then it should be working, right?

yes, your shader set up is fine, and some how adding another layer is causing the issue, and I can't think of any reason yet. deck.gl internally assigns a unique index for each layer and uses it in picking process to identify which layer got picked, and your customizations to ScatterplotLayer shouldn't effect any of this. 

Can you check with two original layer instances (without your customizations) have the same issue or not? If you can provide reproduce steps for me, I can help debug it. (I tried with two scatterplot instances and it works fine).

**Comment by bclinkinbeard** [comment]
Yep, you are correct, they're my custom layers. I will try with normal
scatterplot layers when I get back to my machine but one thing I thought of
is that I'm providing binary data directly as attributes.
instancePositions, etc. Not sure if that could be part of the issue.

**Comment by 1chandu** [comment]
>  I'm providing binary data directly as attributes

If the rendering is fine, then it shouldn't effect `picking/highlighting`. And I am assuming there are no warning or errors in browser console window.

**Comment by Pessimistress** [comment]
@bclinkinbeard 	Any chance you can share a code sample that reproduces this issue? You can use one of our [codepen examples](https://codepen.io/vis-gl/pen/NZrgwZ) as a starting point.

**Comment by bclinkinbeard** [comment]
> I am assuming there are no warning or errors in browser console window.

Correct. The only output is this debug logging, which implies everything is ok.
```
deck: Updated attributes for 7 instances in my-scatterplot-layer-1561408615738 in 3ms
deck: Updated attributes for 7 instances in my-scatterplot-layer-1561408615752 in 0ms
```

I'm pretty stumped here, and I'm finding it difficult to create a test case. I'm using deck on top of Google Maps within an Angular application, but not yet using the GoogleMapsOverlay, so there are a few moving pieces. I took my custom layer/shaders out of the equation and am now just using deck's `ScatterplotLayer` and seeing the same behavior.

**Comment by bclinkinbeard** [comment]
@Pessimistress Just saw your comment. Is there a similar example for ScatterplotLayer?

**Comment by Pessimistress** [comment]
@bclinkinbeard You should be able to simply replace PathLayer with ScatterplotLayer in the example (minus the shader injection). Here's a [sample dataset](https://raw.githubusercontent.com/uber-common/deck.gl-data/master/website/bart-stations.json).

**Comment by bclinkinbeard** [comment]
Thanks @Pessimistress, I confirmed it does work in Codepen so the problem must be in my code. Below is a quick rundown of how I'm doing things, maybe something will stick out as obviously wrong.

As mentioned, I'm building for Angular and Google Maps rather than React and Mapbox. The code is based on what became the `GoogleMapsLayer` but does not yet use that, so I have some manual syncing code in our `draw` method.

```
DeckGLOverlay.deckInstance.setProps({
  layers,
  viewState: {
    zoom: this.fractionalZoom,
    latitude: this.fractionalCenter.lat(),
    longitude: this.fractionalCenter.lng(),
  }
});

// force deck.gl to redraw
if (DeckGLOverlay.deckInstance.layerManager) {
  DeckGLOverlay.deckInstance.animationLoop._setupFrame();
  DeckGLOverlay.deckInstance.animationLoop._updateCallbackData();
  DeckGLOverlay.deckInstance.animationLoop.onRender(
      DeckGLOverlay.deckInstance.animationLoop.animationProps);
}
```

As you can see, `DeckGLOverlay` manages a static instance of deck since there will often be multiple `DeckGLOverlay` subclasses in use at a time.

Basically, each subclass passes its layer descriptors to `super.draw()` which looks for a layer in `DeckGLOverlay.deckInstance.props.layers` with a matching id. If it finds a match it splices in the new descriptor, otherwise it just adds it to the array before running the `setProps` and syncing code above.

Is there anything I should be aware of regarding the manual management of layers? Stuff like `return (<DeckGL {...viewport} layers={[layer]} />);` unfortunately isn't an option for me.

**Comment by Pessimistress** [comment]
Ah that's likely where the problem is. I would try the following:

- Create a new `layers` array on each render instead of mutating the old one
- Use `deck.redraw(true)` to force redraw, instead of manually accessing the `animationLoop`
- It doesn't look like you do, but in case you want to draw into a shared context, you can use [this trick](https://github.com/uber/deck.gl/blob/master/modules/mapbox/src/deck-utils.js#L85)

**Comment by bclinkinbeard** [comment]
> Create a new layers array on each render instead of mutating the old one

Should I also avoid reusing the layer objects I pull out of `deckInstance.props.layers`? I currently don't have an easy way to assign fresh descriptors each render.

**Comment by ibgreen** [comment]
> Should I also avoid reusing the layer objects I pull out of deckInstance.props.layers?

No, you should be able to reuse layer instances (as they are immutable) but mutations of the layers array will likely cause problem as deck.gl optimizes updates with shallow compares.

If you really need it, `layer.clone({...modifiedProps})` does exist.

**Comment by bclinkinbeard** [comment]
Looks like it was my old hacky integration. I've updated to use `GoogleMapsOverlay` and picking works as expected. Thanks for the help and sorry for the goose chase.
